File: nil/cmd/nil/internal/contract/deploy.go
package contract

import (
	"fmt"

	"github.com/NilFoundation/nil/nil/cmd/nil/internal/common"
	"github.com/NilFoundation/nil/nil/cmd/nil/internal/config"
	libcommon "github.com/NilFoundation/nil/nil/common"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/NilFoundation/nil/nil/services/cliservice"
	"github.com/spf13/cobra"
)

func GetDeployCommand(cfg *common.Config) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "deploy [path to file] [args...]",
		Short: "Deploy a smart contract",
		Long:  "Deploy a smart contract with the specified hex-bytecode from stdin or from a file",
		Args:  cobra.MinimumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runDeploy(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	setDeployFlags(cmd)

	return cmd
}

func setDeployFlags(cmd *cobra.Command) {
	cmd.Flags().Var(
		types.NewShardId(&params.shardId, types.BaseShardId),
		shardIdFlag,
		"Specify the shard ID to interact with",
	)

	params.salt = *types.NewUint256(0)
	cmd.Flags().Var(
		&params.salt,
		saltFlag,
		"The salt for deployment message",
	)

	cmd.Flags().StringVar(
		&params.abiPath,
		abiFlag,
		"",
		"The path to the ABI file",
	)

	cmd.Flags().BoolVar(
		&params.noWait,
		noWaitFlag,
		false,
		"Define whether the command should wait for the receipt",
	)

	cmd.Flags().Var(
		&params.feeCredit,
		feeCreditFlag,
		"The deployment fee credit. If  set to 0, it will be estimated automatically",
	)
}

func runDeploy(_ *cobra.Command, cmdArgs []string, cfg *common.Config) error {
	service := cliservice.NewService(common.GetRpcClient(), cfg.PrivateKey)

	var filename string
	var args []string
	if len(cmdArgs) > 0 {
		filename = cmdArgs[0]
		args = cmdArgs[1:]
	}

	bytecode, err := common.ReadBytecode(filename, params.abiPath, args)
	if err != nil {
		return err
	}

	payload := types.BuildDeployPayload(bytecode, libcommon.Hash(params.salt.Bytes32()))

	msgHash, addr, err := service.DeployContractExternal(params.shardId, payload, params.feeCredit)
	if err != nil {
		return err
	}

	if !params.noWait {
		if _, err := service.WaitForReceipt(addr.ShardId(), msgHash); err != nil {
			return err
		}
	}

	if !config.Quiet {
		fmt.Print("Message hash: ")
	}
	fmt.Println(msgHash)

	if !config.Quiet {
		fmt.Print("Contract address: ")
	}
	fmt.Println(addr)

	return nil
}

File: clijs/src/commands/contract/deploy.ts
import {
  type Hex,
  bytesToHex,
  externalDeploymentTransaction,
  waitTillCompleted,
} from "@nilfoundation/niljs";
import { Args, Flags } from "@oclif/core";
import type { Abi } from "abitype";
import { BaseCommand } from "../../base.js";
import { readFile, readJsonFile } from "../../common/utils.js";
import { bigintFlag, hexArg } from "../../types.js";

export default class ContractDeploy extends BaseCommand {
  static override summary = "Deploy a smart contract";
  static override description =
    "Deploy the smart contract with the specified hex-bytecode from stdin or from file";

  static flags = {
    shardId: Flags.integer({
      char: "s",
      description: "Specify the shard ID to interact with",
      required: false,
      default: 1,
    }),
    salt: Flags.integer({
      char: "t",
      description: "The salt for the deploy transaction",
      required: false,
      default: 0,
    }),
    abiPath: Flags.string({
      char: "a",
      description: "The path to the ABI file",
      required: true,
    }),
    feeCredit: bigintFlag({
      char: "f",
      description: "The fee credit for transaction processing",
      required: false,
      default: 10000000000n,
    }),
    noWait: Flags.boolean({
      char: "n",
      description: "Define whether the command should wait for the receipt",
      required: false,
      default: false,
    }),
  };

  static args = {
    filename: hexArg({
      name: "filename",
      required: true,
      description: "Bytecode file name",
    }),
    method: Args.string({
      name: "method",
      required: true,
      description: "Contract methods",
    }),
    args: Args.string({
      name: "args",
      required: false,
      description: "Method arguments",
      multiple: true,
    }),
  };

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  public async run(): Promise<Hex> {
    const { flags, args } = await this.parse(ContractDeploy);

    const abi = readJsonFile<Abi>(flags.abiPath);
    const bytecode = readFile(args.filename);

    if (!this.rpcClient) {
      throw new Error("RPC client is not initialized");
    }

    const chainId = await this.rpcClient.chainId();
    const gasPrice = await this.rpcClient.getGasPrice(flags.shardId);

    const deploymentTransaction = externalDeploymentTransaction(
      {
        salt: BigInt(flags.salt),
        shard: flags.shardId,
        bytecode: bytecode as `0x${string}`,
        abi: abi,
        args: args.args?.split(" ") ?? [],
        feeCredit: flags.feeCredit * gasPrice,
      },
      chainId,
    );
    const address = bytesToHex(deploymentTransaction.to);

    const hash = await deploymentTransaction.send(this.rpcClient);
    if (!flags.noWait) {
      await waitTillCompleted(this.rpcClient, hash);
    }
    this.log(`Transaction hash: ${hash}`);
    return address;
  }
}
